### PR TITLE
[FLINK-17773] Update documentation for new WatermarkGenerator/WatermarkStrategies

### PR DESCRIPTION
**Pull request by aljoscha**
## What is the purpose of the change

This brings the documentation up-to-date with the latest changes. This is not a simple find-replace job, I also restructured/updated the documentation and removed so sections that contain outdated advice and added new sections.

It's probably easiest to not look at the diff but to build the docs locally and check out the event-time section.

@sjwiesman @alpinegizmo Could you please review this?

fyi @StephanEwen

### CODE REVIEW

File: docs/dev/event_time.md [hunk 1]
```diff
@@ -34,30 +34,22 @@ For information about how to use time in Flink programs refer to
...
 final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
-env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime);
-
-// alternatively:
-// env.setStreamTimeCharacteristic(TimeCharacteristic.IngestionTime);
-// env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
 
-DataStream<MyEvent> stream = env.addSource(new FlinkKafkaConsumer010<MyEvent>(topic, schema, props));
```
Comment (sjwiesman) [inline docs/dev/event_time.md:60]: +1

File: docs/dev/event_timestamps_watermarks.md [hunk 2]
```diff
@@ -22,115 +22,104 @@ specific language governing permissions and limitations
...
-{% endhighlight %}
-</div>
-</div>
-
-## Assigning Timestamps
-
-In order to work with *event time*, Flink needs to know the events' *timestamps*, meaning each element in the
-stream needs to have its event timestamp *assigned*. This is usually done by accessing/extracting the
-timestamp from some field in the element.
+In order to work with *event time*, Flink needs to know the events'
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:35]:
```suggestion
In order to work with *event time*, Flink needs to know the events
```

File: docs/dev/event_timestamps_watermarks.md [hunk 3]
```diff
@@ -22,115 +22,104 @@ specific language governing permissions and limitations
...
+*timestamps*, meaning each element in the stream needs to have its event
+timestamp *assigned*. This is usually done by accessing/extracting the
+timestamp from some field in the element by using a `TimestampAssigner`.
 
-Timestamp assignment goes hand-in-hand with generating watermarks, which tell the system about
-progress in event time.
+Timestamp assignment goes hand-in-hand with generating watermarks, which tell
+the system about progress in event time. You can configure this by speciying a
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:41]:
```suggestion
the system about progress in event time. You can configure this by specifying a
```

File: docs/dev/event_timestamps_watermarks.md [hunk 4]
```diff
@@ -22,115 +22,104 @@ specific language governing permissions and limitations
...
+`WatermarkGenerator`.
 
-There are two ways to assign timestamps and generate watermarks:
+The Flink API expects both `TimestampAssigner` and `WatermarkGenerator` to be
+bundled in a `WatermarkStrategy`. As you will see below, you usually don't have
+to do this by hand because you will mostly use the convenience helper
+`WatermarkStrategies`. Here is the interface for completess' sake:
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:47]: I think it always makes sense to let users know there is an easy option first and then outline the low level interface. Its less scary 👻 
```suggestion
The Flink API expects a `WatermarkStrategy` that contains both a `TimestampAssigner` and `WatermarkGenerator`.
A number of common strategies out of the box, available in the `WatermarkStrategies` helper, but users can also build their own strategies when required. 
Here is the interface for completeness' sake:
```

File: docs/dev/event_timestamps_watermarks.md [hunk 5]
```diff
@@ -175,18 +164,103 @@ withTimestampsAndWatermarks
...
 
-`AssignerWithPeriodicWatermarks` assigns timestamps and generates watermarks periodically (possibly depending
-on the stream elements, or purely based on processing time).
+### Writing a Periodic WatermarkGenerator
 
-The interval (every *n* milliseconds) in which the watermark will be generated is defined via
-`ExecutionConfig.setAutoWatermarkInterval(...)`. The assigner's `getCurrentWatermark()` method will be
-called each time, and a new watermark will be emitted if the returned watermark is non-null and larger than the previous
-watermark.
+As mentioned above, a periodic generator observes stream events and generates
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:248]: You use the phrase "as mentioned above" alot. 
```suggestion
A periodic generator observes stream events and generates
```

File: docs/dev/event_timestamps_watermarks.md [hunk 6]
```diff
@@ -270,105 +340,115 @@ class TimeLagWatermarkGenerator extends AssignerWithPeriodicWatermarks[MyEvent]
...
 </div>
 
-#### **With Punctuated Watermarks**
+### Writing a Punctuated WatermarkGenerator
 
-To generate watermarks whenever a certain event indicates that a new watermark might be generated, use
-`AssignerWithPunctuatedWatermarks`. For this class Flink will first call the `extractTimestamp(...)` method
-to assign the element a timestamp, and then immediately call the
-`checkAndGetNextWatermark(...)` method on that element.
+As mentioned above, a punctuated watermark generator will observe the stream of
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:357]:
```suggestion
A punctuated watermark generator will observe the stream of
```

File: docs/dev/event_timestamps_watermarks.md [hunk 7]
```diff
@@ -377,4 +457,77 @@ val stream: DataStream[MyType] = env.addSource(kafkaSource)
...
+public void run(SourceContext<MyType> ctx) throws Exception {
+    while (/* condition */) {
+        MyType next = getNext();
+        ctx.collectWithTimestamp(next, next.getEventTimestamp());
+
+        if (next.hasWatermarkTime()) {
+            ctx.emitWatermark(new Watermark(next.getWatermarkTime()));
+        }
+    }
+}
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:487]: Should this use the new source interface?

File: docs/dev/event_timestamps_watermarks.md [hunk 8]
```diff
@@ -377,4 +457,77 @@ val stream: DataStream[MyType] = env.addSource(kafkaSource)
...
+        if (next.hasWatermarkTime) {
+            ctx.emitWatermark(new Watermark(next.getWatermarkTime))
+        }
+    }
+}
+{% endhighlight %}
+</div>
+</div>
+
+## How Operators Are Processing Watermarks
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:506]:
```suggestion
## How Operators Process Watermarks
```

File: docs/dev/event_timestamps_watermarks.md [hunk 9]
```diff
@@ -377,4 +457,77 @@ val stream: DataStream[MyType] = env.addSource(kafkaSource)
...
+
+As a general rule, operators are required to completely process a given
+watermark before forwarding it downstream. For example, `WindowOperator` will
+first evaluate which windows should be fired, and only after producing all of
```
Comment (sjwiesman) [inline docs/dev/event_timestamps_watermarks.md:510]:
```suggestion
first evaluate all windows that should be fired, and only after producing all of
```

File: docs/dev/event_timestamp_extractors.md [hunk 10]
```diff
@@ -25,83 +25,80 @@ under the License.
...
+them. Apart from their out-of-the-box functionality, their implementation can
+serve as an example for custom implementations.
 
-### **Assigners with ascending timestamps**
+## Monotonously Increasing Timestamps
 
-The simplest special case for *periodic* watermark generation is the case where timestamps seen by a given source task
-occur in ascending order. In that case, the current timestamp can always act as a watermark, because no earlier timestamps will
-arrive.
+The simplest special case for *periodic* watermark generation is the case where
```
Comment (sjwiesman) [inline docs/dev/event_timestamp_extractors.md:41]: You use the word case three times in two sentences.
```suggestion
The simplest special case for *periodic* watermark generation is the when
```

File: docs/dev/event_timestamps_watermarks.md (hunk 7 above)
Comment (aljoscha) [reply docs/dev/event_timestamps_watermarks.md:486]: For this one, I just copied the existing example. I might update to the new Source interface as a follow-up, but that's very different. Maybe we shouldn't even mention this here?

File: docs/dev/event_timestamp_extractors.md (hunk 10 above)
Comment (aljoscha) [reply docs/dev/event_timestamp_extractors.md:41]: In my defense: this is existing copy that I just reused... 😅

**Comment by aljoscha** [comment]
Thanks for the thorough review!

File: docs/dev/event_timestamps_watermarks.md (hunk 7 above)
Comment (sjwiesman) [reply docs/dev/event_timestamps_watermarks.md:486]: Yeah, let’s remove it. How many people are really building custom sources anyway.

**Comment by sjwiesman** [comment]
@aljoscha let’s remove the custom source stuff and the +1 to merge

**Comment by aljoscha** [comment]
👌I'm merging

**Comment by aljoscha** [comment]
@sjwiesman do you want to jump on https://issues.apache.org/jira/browse/FLINK-17886

**Pull request closed without merging**